File: src/main/java/net/casualuhc/uhcmod/managers/WorldBorderManager.java
package net.casualuhc.uhcmod.managers;

import net.casualuhc.arcade.border.CustomBorder;
import net.casualuhc.arcade.events.EventHandler;
import net.casualuhc.arcade.events.border.CustomBorderFinishEvent;
import net.casualuhc.arcade.events.server.ServerTickEvent;
import net.casualuhc.arcade.scheduler.MinecraftTimeUnit;
import net.casualuhc.uhcmod.UHCMod;
import net.casualuhc.uhcmod.events.uhc.UHCBorderCompleteEvent;
import net.casualuhc.uhcmod.events.uhc.UHCGracePeriodEndEvent;
import net.casualuhc.uhcmod.utils.gamesettings.GameSettings;
import net.casualuhc.uhcmod.utils.uhc.Phase;
import net.minecraft.server.MinecraftServer;
import net.minecraft.world.border.WorldBorder;

public class WorldBorderManager {
	private static final MinecraftServer SERVER = UHCMod.SERVER;
	public static final CustomBorder GLOBAL_BORDER = new CustomBorder();
	public static final int PORTAL_ESCAPE_TIME_SECONDS = 30;

	public static void startWorldBorders() {
		double size = UHCMod.SERVER.getOverworld().getWorldBorder().getSize();
		Stage stage = Stage.getStage(size);
		if (stage == null) {
			stage = Stage.FIRST;
			moveWorldBorders(Stage.FIRST.getStartSize(), 0);
		}

		GameSettings.WORLD_BORDER_STAGE.setValueQuietly(stage);
		if (stage == Stage.END) {
			EventHandler.broadcast(new UHCBorderCompleteEvent());
			return;
		}

		moveWorldBorders(stage.getEndSize(), stage.getTime(size));
	}

	public static void moveWorldBorders(double newSize, long time) {
		long modifiedTime = (long) (time * GameSettings.WORLD_BORDER_SPEED.getValue());
		SERVER.getWorlds().forEach(world -> {
			WorldBorder border = world.getWorldBorder();
			if (modifiedTime != 0) {
				border.interpolateSize(border.getSize(), newSize, modifiedTime * 1000);
				return;
			}
			border.setSize(newSize);
		});
	}

	public static void noop() { }

	static {
		EventHandler.register(CustomBorderFinishEvent.class, event -> {
			UHCMod.LOGGER.info("Finished world border: {}", GameSettings.WORLD_BORDER_STAGE.getValue());

			Stage nextStage = GameSettings.WORLD_BORDER_STAGE.getValue().getNextStage();
			if (nextStage == null || !UHCManager.isPhase(Phase.ACTIVE)) {
				return;
			}

			GameSettings.WORLD_BORDER_STAGE.setValueQuietly(nextStage);
			if (nextStage == Stage.END) {
				EventHandler.broadcast(new UHCBorderCompleteEvent());
				return;
			}

			UHCManager.schedulePhaseTask(10, MinecraftTimeUnit.Seconds, () -> {
				double size = UHCMod.SERVER.getOverworld().getWorldBorder().getSize();
				moveWorldBorders(nextStage.getEndSize(), nextStage.getTime(size));
			});
		});
		EventHandler.register(UHCGracePeriodEndEvent.class, event -> {
			startWorldBorders();
		});
		EventHandler.register(ServerTickEvent.class, event -> {
			GLOBAL_BORDER.update();
		});
	}

	public enum Stage {
		FIRST(6128, 3064, 1800),
		SECOND(3064, 1532),
		THIRD(1532, 766, 900),
		FOURTH(766, 383, 800),
		FIFTH(383, 180, 700),
		SIX(180, 50, 500),
		FINAL(50, 20, 200),
		END(20, 0, 0);

		private final long startSize;
		private final long endSize;
		private final long time;

		Stage(long startSize, long endSize, long time) {
			this.startSize = startSize;
			this.endSize = endSize;
			this.time = time;
		}

		Stage(long startSize, long endSize) {
			this(startSize, endSize, startSize - endSize);
		}

		public float getStartSize() {
			return this.startSize;
		}

		public float getEndSize() {
			return this.endSize;
		}

		public long getTime(double size) {
			double blockPerTime = (this.startSize - this.endSize) / (double) this.time;
			return (long) (((long) size - this.endSize) / blockPerTime);
		}

		public Stage getNextStage() {
			int next = this.ordinal() + 1;
			Stage[] values = Stage.values();

			if (next < values.length) {
				return values[next];
			}
			return END;
		}

		public static Stage getStage(double size) {
			if (size <= FINAL.endSize) {
				return END;
			}
			for (Stage stage : Stage.values()) {
				if (size <= stage.startSize && size > stage.endSize) {
					return stage;
				}
			}
			return null;
		}
	}
}

File: src/main/kotlin/net/casualuhc/uhcmod/managers/WorldBorderManager.kt
package net.casualuhc.uhcmod.managers

import net.casualuhc.arcade.border.MultiLevelBorderListener
import net.casualuhc.arcade.border.MultiLevelBorderTracker
import net.casualuhc.arcade.border.TrackedBorder
import net.casualuhc.arcade.events.GlobalEventHandler
import net.casualuhc.arcade.events.server.ServerLoadedEvent
import net.casualuhc.arcade.scheduler.MinecraftTimeUnit.Seconds
import net.casualuhc.arcade.utils.BorderUtils
import net.casualuhc.arcade.utils.LevelUtils
import net.casualuhc.uhcmod.UHCMod
import net.casualuhc.uhcmod.events.uhc.UHCBorderCompleteEvent
import net.casualuhc.uhcmod.events.uhc.UHCGracePeriodEndEvent
import net.casualuhc.uhcmod.settings.GameSettings
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.Level

object WorldBorderManager: MultiLevelBorderListener {
    private val tracker = MultiLevelBorderTracker()

    const val PORTAL_ESCAPE_TIME_SECONDS = 30

    var stage: Stage
        get() = GameSettings.WORLD_BORDER_STAGE.value
        private set(value) = GameSettings.WORLD_BORDER_STAGE.setValueQuietly(value)
    val multiplier: Long
        get() = GameSettings.WORLD_BORDER_TIME.value

    fun startWorldBorders() {
        if (this.stage == Stage.END) {
            GlobalEventHandler.broadcast(UHCBorderCompleteEvent())
            return
        }
        this.moveWorldBorders(this.stage)
    }

    fun moveWorldBorders(stage: Stage, size: Size = Size.END, instant: Boolean = false) {
        for ((border, level) in this.tracker.getAllTracking()) {
            this.moveWorldBorder(border, level, stage, size, instant)
        }
    }

    fun pauseWorldBorders() {
        for ((border, _) in this.tracker.getAllTracking()) {
            this.moveWorldBorder(border, border.size)
        }
    }

    override fun onAllBordersComplete(borders: Map<TrackedBorder, ServerLevel>) {
        UHCMod.logger.info("Finished world border stage: ${this.stage}")
        if (!UHCManager.isActivePhase()) {
            return
        }
        val next = this.stage.getNextStage()
        this.stage = next

        if (next == Stage.END) {
            GlobalEventHandler.broadcast(UHCBorderCompleteEvent())
            return
        }
        super.onAllBordersComplete(borders)
    }

    override fun onAllBordersComplete(border: TrackedBorder, level: ServerLevel) {
        // We don't shrink past the fifth stage in the end because
        // otherwise it becomes impossible to enter the end dimension
        if (level.dimension() == Level.END && this.stage > Stage.FIFTH) {
            return
        }

        UHCManager.schedulePhaseTask(10, Seconds) {
            this.moveWorldBorder(border, level, this.stage, Size.END)
        }
    }

    internal fun registerEvents() {
        GlobalEventHandler.register<ServerLoadedEvent> { this.onServerLoadedEvent() }
        GlobalEventHandler.register<UHCGracePeriodEndEvent> { this.startWorldBorders() }
    }

    private fun moveWorldBorder(border: TrackedBorder, level: Level, stage: Stage, size: Size, instant: Boolean = false) {
        val dest = if (size == Size.END) stage.getEndSizeFor(level) else stage.getStartSizeFor(level)
        val time = if (instant) -1.0 else stage.getRemainingTimeAsPercent(border.size, level)
        this.moveWorldBorder(border, dest, time)
    }

    private fun moveWorldBorder(border: TrackedBorder, newSize: Double, percent: Double = -1.0) {
        val seconds = (percent * this.multiplier).toLong()
        if (seconds > 0) {
            border.lerpSizeBetween(border.size, newSize, seconds * 1000L)
            return
        }
        border.size = newSize
    }

    private fun onServerLoadedEvent() {
        this.tracker.addLevelBorder(LevelUtils.overworld())
        this.tracker.addLevelBorder(LevelUtils.nether())
        this.tracker.addLevelBorder(LevelUtils.end())

        BorderUtils.isolateWorldBorders()
    }

    enum class Size {
        START, END
    }

    enum class Stage(
        private val startSize: Double,
        private val endSize: Double,
        private val weight: Double
    ) {
        FIRST(6128.0, 3064.0, 18.0),
        SECOND(FIRST.endSize, 1500.0, 15.0),
        THIRD(SECOND.endSize, 780.0, 9.0),
        FOURTH(THIRD.endSize, 400.0, 8.0),
        FIFTH(FOURTH.endSize, 220.0, 7.0),
        SIX(FIFTH.endSize, 50.0, 5.0),
        FINAL(SIX.endSize, 20.0, 2.0),
        END(FINAL.endSize, 0.0, 0.0);

        fun getRemainingTimeAsPercent(size: Double, level: Level): Double {
            val remainingSize = size - this.getEndSizeFor(level)
            val totalSize = this.getStartSizeFor(level) - this.getEndSizeFor(level)
            return (remainingSize / totalSize) * (this.weight / TOTAL_WEIGHT)
        }

        fun getNextStage(): Stage {
            val next = this.ordinal + 1
            val values = values()
            return if (next < values.size) values[next] else END
        }

        fun getStartSizeFor(level: Level): Double {
            return adjustSize(this.startSize, level)
        }

        fun getEndSizeFor(level: Level): Double {
            return adjustSize(this.endSize, level)
        }

        companion object {
            val TOTAL_WEIGHT = values().sumOf { it.weight }

            fun getStage(size: Double, level: Level): Stage? {
                val adjusted = reverseAdjustSize(size, level)
                if (adjusted <= FINAL.endSize) {
                    return END
                }
                for (stage in values()) {
                    if (adjusted <= stage.startSize && adjusted > stage.endSize) {
                        return stage
                    }
                }
                return null
            }

            fun adjustSize(size: Double, level: Level): Double {
                return size / level.dimensionType().coordinateScale
            }

            fun reverseAdjustSize(size: Double, level: Level): Double {
                return size * level.dimensionType().coordinateScale
            }
        }
    }
}
